src/main/java/se/liu/ida/hefquin/engine/data/impl/VocabularyMappingImpl.java

```diff
@@ -1,62 +1,326 @@
+import org.apache.jena.graph.Node;
+import org.apache.jena.graph.Triple;
 import org.apache.jena.rdf.model.Model;
 import org.apache.jena.riot.RDFDataMgr;
 import org.apache.jena.sparql.graph.GraphFactory;
+import org.apache.jena.vocabulary.OWL;
+import org.apache.jena.vocabulary.RDF;
+import org.apache.jena.vocabulary.RDFS;
 
-import se.liu.ida.hefquin.engine.data.Triple;
 import se.liu.ida.hefquin.engine.data.VocabularyMapping;
+import se.liu.ida.hefquin.engine.query.BGP;
 import se.liu.ida.hefquin.engine.query.SPARQLGraphPattern;
+import se.liu.ida.hefquin.engine.query.SPARQLUnionPattern;
 import se.liu.ida.hefquin.engine.query.TriplePattern;
+import se.liu.ida.hefquin.engine.query.impl.BGPImpl;
+import se.liu.ida.hefquin.engine.query.impl.SPARQLGroupPatternImpl;
+import se.liu.ida.hefquin.engine.query.impl.SPARQLUnionPatternImpl;
+import se.liu.ida.hefquin.engine.query.impl.TriplePatternImpl;
 
-public class VocabularyMappingImpl implements VocabularyMapping{
-	
+public class VocabularyMappingImpl implements VocabularyMapping
+{
 	protected final Graph vocabularyMapping;
 	
 	public VocabularyMappingImpl() {
 		vocabularyMapping = GraphFactory.createDefaultGraph();
 	}
 	
-	
 	// Source: https://jena.apache.org/documentation/io/rdf-input.html
 	public VocabularyMappingImpl(final String rdfFile) {
-		Model mappingModel = RDFDataMgr.loadModel(rdfFile); //.nt file for N-Triple
+		final Model mappingModel = RDFDataMgr.loadModel(rdfFile); //.nt file for N-Triple
 		vocabularyMapping = mappingModel.getGraph();
 	}
 	
 	public VocabularyMappingImpl(final Set<Triple> triples) {
-		Graph defaultGraph = GraphFactory.createDefaultGraph();
-		final Iterator<Triple> i = triples.iterator();
-		while(i.hasNext()) {
-			final Triple t = i.next();
-			defaultGraph.add(t.asJenaTriple());
+		vocabularyMapping = GraphFactory.createDefaultGraph();
+		for ( final Triple t : triples ) {
+			vocabularyMapping.add(t);
 		}
-		vocabularyMapping = defaultGraph;
 	}
 
 	@Override
 	public SPARQLGraphPattern translateTriplePattern(final TriplePattern tp) {
-		return null;
+		final List<SPARQLGraphPattern> objectTranslation = new ArrayList<>();
+		for(final TriplePattern i : translateSubject(tp)) {
+			objectTranslation.add(translateObject(i));
+		}
+		
+		final List<SPARQLGraphPattern> predicateTranslation = new ArrayList<>();
+		for(final SPARQLGraphPattern j : objectTranslation) {
+			
+			if(j instanceof SPARQLUnionPattern) {
+				
+				final SPARQLUnionPatternImpl union = new SPARQLUnionPatternImpl();
+				for(final SPARQLGraphPattern k: ((SPARQLUnionPattern) j).getSubPatterns()) {		
+					if(k instanceof TriplePattern) {
+						union.addSubPattern(translatePredicate((TriplePattern) k));
+					} else if (k instanceof BGP) {
+						for(final TriplePattern l : ((BGP) k).getTriplePatterns()) {
+// TODO: I think that this part here is incorrect, please double check
+							union.addSubPattern(translatePredicate(l));
+						}
+					} else {
```

Comment:
@JulietteVV I have cleaned up the code a bit more. In this context, I have come across this part, which I think it is incorrect. Can you please double check it again. The issue that I mean here is that you are essentially turning a BGP into a union; in other words, the fact the `k` was a BGP gets completely lost.

Reply:
Yes I see now that I forgot to change this, it should be fixed now!

Reply:
Yes, that fixes it. Thanks!